query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java

```diff
@@ -333,81 +313,80 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteM
-      }
-   }
-
-   private void purgeIndex(TransactionContext transactionContext, Class<?> entityType) {
-      IndexedTypeIdentifier type = new PojoIndexedTypeIdentifier(entityType);
-      if (searchFactory.getIndexBindings().containsKey(type)) {
-         performSearchWork(searchWorkCreator.createPerEntityTypeWork(type, WorkType.PURGE_ALL), transactionContext);
-      }
-   }
-
-   private void purgeAllIndexes(TransactionContext transactionContext) {
-      for (IndexedTypeIdentifier type : searchFactory.getIndexBindings().keySet()) {
-         performSearchWork(searchWorkCreator.createPerEntityTypeWork(type, WorkType.PURGE_ALL), transactionContext);
-      }
+   void removeFromIndexes(Object key, int segment) {
+      performSearchWorks((indexer) -> indexer.purge(keyToString(key, segment), segment+""));
    }
 
    // Method that will be called when data needs to be removed from Lucene.
-   private void removeFromIndexes(Object value, Object key, TransactionContext transactionContext, int segment) {
-      performSearchWork(value, keyToString(key, segment), WorkType.DELETE, transactionContext);
-   }
-
-   private void updateIndexes(boolean usingSkipIndexCleanupFlag, Object value, Object key, TransactionContext transactionContext, int segment) {
-      // Note: it's generally unsafe to assume there is no previous entry to cleanup: always use UPDATE
-      // unless the specific flag is allowing this.
-      performSearchWork(value, keyToString(key, segment), usingSkipIndexCleanupFlag ? WorkType.ADD : WorkType.UPDATE, transactionContext);
+   private void removeFromIndexes(Object value, Object key, int segment) {
+      performSearchWorks((indexer) -> indexer.delete(keyToString(key, segment), value));
    }
 
-   private void performSearchWork(Object value, Serializable id, WorkType workType, TransactionContext transactionContext) {
-      if (value == null) throw new NullPointerException("Cannot handle a null value!");
-      performSearchWork(searchWorkCreator.createPerEntityWork(value, id, workType), transactionContext);
+   private void updateIndexes(boolean usingSkipIndexCleanupFlag, Object value, Object key, int segment) {
+      performSearchWorks((indexer) -> {
+         // Note: it's generally unsafe to assume there is no previous entry to cleanup: always use UPDATE
+         // unless the specific flag is allowing this.
+         if (usingSkipIndexCleanupFlag) {
+            return indexer.add(keyToString(key, segment), value);
+         } else {
+            return indexer.addOrUpdate(keyToString(key, segment), value);
+         }
+      });
    }
 
-   private void performSearchWork(Work work, TransactionContext transactionContext) {
-      if (work != null) {
-         Worker worker = searchFactory.getWorker();
-         worker.performWork(work, transactionContext);
-      }
+   private void performSearchWorks(Function<SearchIndexer, CompletableFuture<?>> works) {
```

Comment:
@fax4ever This lambda does not seem to bring much. Can't you just call `searchMappingHolder.getSearchMapping().getSearchIndexer()` from both `updateIndexes` and `removeFromIndexes`?

Reply:
Good catch @yrodiere. No reason for keeping this lamda... I'm removing it

Reply:
done... you can find the fix on this branch